### DISCUSSION TITLE
Does WASM have a role to play for cross-platform m-ld engines?

### DISCUSSION THREAD

**Discussion by mcalligator** [m-ld/m-ld-spec#95] [Ideas]
[This paper from 2019](https://arxiv.org/ftp/arxiv/papers/2010/2010.07115.pdf) explores the potential for WASM in serverless computing.  Whilst not *directly* relevant to m-ld, it was a reference for [this article](https://blog.colinbreck.com/why-am-i-excited-about-webassembly/) about using WASM for IoT, in the light of its:

1. Sandbox approach to security (echoes of JVM there);
2. Portability across different OS platforms; and
3. Small footprint (aiding greater performance).

In terms of challenges, WASM isn't perceived to be fast in the browser, but the paper demonstrates that some WASM runtime implementations are much quicker for the things tested than their Docker counterparts.  Another (current) drawback is the relative paucity of higher-level language support in WASM (C++, Rust, Python, and perhaps others).  However, WASM's increasing maturity makes it worth investigating further as a means of taking m-ld cross-platform more quickly.  This discussion is for exploring that.

**Comment by gsvarovsky** [comment]
A key desirable quality of m-ld is that it should _stay out of the way_. That is, in terms of:
- Developer cognitive overhead,
- Runtime overhead, and
- Operational overhead

it should have the minimum possible impact, compared to writing the app against a local, unshared, in-memory object model. For example, using m-ld in an application should have a negligible impact on CPU utilisation beyond the app itself.

With such a goal, its clear that using a dedicated runtime _for m-ld itself_ (separately to the application) would have to be strongly justified, as it could negatively impact deployment complexity, and potentially, runtime complexity, when crossing boundaries – this is the source of performance concerns with WASM in a browser; not the raw speed of the internal runtime. So, how much does m-ld need to cross boundaries?

The article focusses on 'pure functional' components, because they can be sandboxed. From the app's point of view, m-ld is decidedly not pure functional – it has side-effects on local storage, and of course other clones via the network. The core algorithm can be cajoled into feeling pure for the sake of argument by taking a more holistic view, but we can sidestep that because really, the question is whether it can work effectively in a sandbox, and the answer to that is no. Most of the overhead of using m-ld is not in the core algorithm but in the interaction with persistence and with the network. Anything that slows those things down is going to swamp any positive effect on the speed of the core algorithm.

Can we avoid this by pulling other stuff into the WASM sandbox, so no boundary need be crossed? In the extreme, if you have all the clones, with their persistence, the network, and the app, all within the same WASM runtime, you just have a very nerdy snowglobe. However, some apps might have a use for a high-performance data processing clone whose only interface is the network – to implement some wild AI maybe, and have the results visible to other domain participants. We can probably envisage other scenarios for which a high-performance sandboxed runtime would be interesting.

In general though, I think the best model for a m-ld engine is as a local library, in the native runtime of the app; with which it shares privileges for storage and network access. While that suggests re-implementing m-ld in all languages, this can be mitigated by using existing native interfaces. For example, a Rust or C implementation can be compiled for just about any platform and called with almost no overhead from any language. In all cases there'll still be swings and roundabouts to consider, for sure.
